library/std/src/sys/windows/rand.rs

```diff
@@ -34,35 +33,77 @@ use crate::sys::c;
 /// [`HashMap`]: crate::collections::HashMap
 /// [`RandomState`]: crate::collections::hash_map::RandomState
 pub fn hashmap_random_keys() -> (u64, u64) {
-    let mut v = (0, 0);
-    let ret = unsafe {
-        let size = mem::size_of_val(&v).try_into().unwrap();
-        c::BCryptGenRandom(
-            // BCRYPT_RNG_ALG_HANDLE is only supported in Windows 10+.
-            // So for Windows 8.1 and Windows 7 we'll need a fallback when this fails.
-            ptr::invalid_mut(c::BCRYPT_RNG_ALG_HANDLE),
-            ptr::addr_of_mut!(v).cast(),
-            size,
-            0,
-        )
-    };
-    if ret != 0 { fallback_rng() } else { v }
+    Rng::open().and_then(|rng| rng.gen_random_keys()).unwrap_or_else(fallback_rng)
+}
+
+struct Rng(c::BCRYPT_ALG_HANDLE);
+impl Rng {
+    // Open a handle to the RNG algorithm.
+    fn open() -> Result<Self, c::NTSTATUS> {
+        use crate::sync::atomic::AtomicIsize;
+        use crate::sync::atomic::Ordering::{Acquire, Release};
+
+        // An atomic is used so we don't need to reopen the handle every time.
+        static HANDLE: AtomicIsize = AtomicIsize::new(0);
+
+        let mut handle = HANDLE.load(Acquire);
+        if handle == 0 {
+            let status = unsafe {
+                c::BCryptOpenAlgorithmProvider(
+                    &mut handle,
+                    c::BCRYPT_RNG_ALGORITHM.as_ptr(),
+                    ptr::null(),
+                    0,
+                )
+            };
+            if c::nt_success(status) {
+                // If another thread opens a handle first then use that handle instead.
+                let result = HANDLE.compare_exchange(0, handle, Release, Acquire);
+                if let Err(previous_handle) = result {
+                    // Close our handle and return the previous one.
+                    unsafe { c::BCryptCloseAlgorithmProvider(handle, 0) };
+                    handle = previous_handle;
+                }
+                Ok(Self(handle))
+            } else {
+                Err(status)
```

Comment:
If this fails, we don't remember that fact, so we'll keep retrying `BCryptOpenAlgorithmProvider` every time.

Perhaps we should store `INVALID_HANDLE` (or some other sentinel) in the atomic in this case to indicate "opening failed, don't bother trying repeatedly", and return an Err if we see it.

Reply:
I'm hopeful that case will now never be reached but I guess it does make sense to avoid having to try again if it is. I wouldn't use the `INVALID_HANDLE` constant though as that only has a meaning in the context of certain filesystem APIs. I mean, even if we use the same actual value I'd want to give it a different name to avoid conflating the two things.

Reply:
I used a local constant instead, which I'm happier about. Also I wasn't sure what to return as the error so I picked something that seems reasonable enough to me.